### DISCUSSION TITLE
Ignore upgrades based on source versions?

### DISCUSSION THREAD

**Discussion by maxime-michel** [renovatebot/renovate#19421] [Request Help] [answered]
### What would you like Renovate to be able to do?

This use case is marginal, but we have seen it in other codebases and use it to this day because it's handy. We call it 'permanent SNAPSHOT' development. Imagine you have two modules, Maven in our case, with module B @ 1.2.3-SNAPSHOT depending on module A @ 4.5.6-SNAPSHOT. Now, given a frequent enough release cycle, what will happen is:

Module A 4.5.6 gets released, module B is now depending on a SNAPSHOT version that's still around in developer local repositories, but that doesn't exist anymore. Also, the version in module B must either be updated to 4.5.6 or 4.5.7-SNAPSHOT, and this should be pulled by any developer working on the project.

Given a complex enough codebase and enough developers, people forget to either make those updates and/or pull to get the latest status, preventing them from working directly on an up-to-date local environment.

What we do with permanent SNAPSHOTs instead is to develop all modules with a SNAPSHOT version that doesn't include the particular incremental, e.g. module B is now 1.2-SNAPSHOT and module A 4.5-SNAPSHOT. This means that the only moment where the module A dependency in module B is at 4.5.6 is during the release tag creation. Before the release & as soon as it's performed, that is, 99% of the time, module B depends on module A @ 4.5-SNAPSHOT. This allows for stabler development environments.

Now, going back to Renovate, the problem is that it will prompt us to upgrade module's B dependency to module A 4.5-SNAPSHOT to 4.5.6 when it gets released, then 4.5.7 when it gets released, etc. Hence, in our case we would rather be able to ignore any upgrade where the source version matches `X.Y-SNAPSHOT`. But, according to what I've seen, it is only possible to filter target versions.

### If you have any ideas on how this should be implemented, please tell us here.

Introduce a new parameter called `allowedSourceVersions`?

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by viceice** [comment, marked as answer]
use `matchCurrentValue`[^1] or `matchCurrentVersion`[^2]

[^1]: https://docs.renovatebot.com/configuration-options/#matchcurrentvalue
[^2]: https://docs.renovatebot.com/configuration-options/#matchcurrentversion

**Reply by maxime-michel** [reply]
Thanks, that's interesting for sure, but contrary to what I described (filtering based on the source value), this wouldn't allow me to filter out the X.Y-SNAPSHOT generally. It does allow me to go through the dozens of modules that we have, and to exclude 4.5.x incrementals for module A, 1.2.x for module B, etc. But I must maintain that list manually, which I'm not going to do because too inconvenient.

**Reply by viceice** [reply]
currentValue === source value === the value renovate has extracted. So you can use that to filter when you combine a `matchCurrentValue`  regex with a `allowedVersions`[^1] action

[^1]: https://docs.renovatebot.com/configuration-options/#allowedversions

**Reply by viceice** [reply]
something like this
```json
    {
      "matchCurrentValue": "/^\\d+\\.\\d+-SNAPSHOT$/",
      "allowedVersions": "/^\\d+\\.\\d+-SNAPSHOT$/"
    }
```

**Reply by maxime-michel** [reply]
That's awesome, thank you, let me try that. I wasn't aware something like this would be possible.
